finality-aleph/src/validator_network/manager/mod.rs

```diff
@@ -30,191 +41,165 @@ impl Display for SendError {
     }
 }
 
-/// Possible results of adding connections.
-#[derive(Debug, PartialEq, Eq)]
-pub enum AddResult {
-    /// We do not want to maintain a connection with this peer.
-    Uninterested,
-    /// Connection added.
-    Added,
-    /// Old connection replaced with new one.
-    Replaced,
-}
-
-/// Network component responsible for holding the list of peers that we
-/// want to connect to, and managing the established connections.
-pub struct Manager<A: Data, D: Data> {
-    addresses: HashMap<AuthorityId, Vec<A>>,
-    outgoing: HashMap<AuthorityId, mpsc::UnboundedSender<D>>,
-    incoming: HashMap<AuthorityId, mpsc::UnboundedSender<D>>,
-}
-
 struct ManagerStatus {
-    wanted_peers: usize,
-    both_ways_peers: HashSet<AuthorityId>,
     outgoing_peers: HashSet<AuthorityId>,
+    missing_outgoing: HashSet<AuthorityId>,
     incoming_peers: HashSet<AuthorityId>,
-    missing_peers: HashSet<AuthorityId>,
+    missing_incoming: HashSet<AuthorityId>,
 }
 
 impl ManagerStatus {
     fn new<A: Data, D: Data>(manager: &Manager<A, D>) -> Self {
-        let incoming: HashSet<_> = manager
-            .incoming
-            .iter()
-            .filter(|(_, exit)| !exit.is_closed())
-            .map(|(k, _)| k.clone())
-            .collect();
-        let outgoing: HashSet<_> = manager
-            .outgoing
-            .iter()
-            .filter(|(_, exit)| !exit.is_closed())
-            .map(|(k, _)| k.clone())
-            .collect();
-
-        let both_ways = incoming.intersection(&outgoing).cloned().collect();
-        let incoming: HashSet<_> = incoming.difference(&both_ways).cloned().collect();
-        let outgoing: HashSet<_> = outgoing.difference(&both_ways).cloned().collect();
-        let missing = manager
-            .addresses
-            .keys()
-            .filter(|a| !both_ways.contains(a) && !incoming.contains(a) && !outgoing.contains(a))
-            .cloned()
-            .collect();
-
+        let mut incoming_peers = HashSet::new();
+        let mut missing_incoming = HashSet::new();
+        let mut outgoing_peers = HashSet::new();
+        let mut missing_outgoing = HashSet::new();
+
+        for peer in manager.wanted.incoming_peers() {
+            match manager.active_connection(peer) {
+                true => incoming_peers.insert(peer.clone()),
+                false => missing_incoming.insert(peer.clone()),
+            };
+        }
+        for peer in manager.wanted.outgoing_peers() {
```

Comment:
I'm not sure yet, but what about `manager.have`? Is it subset of `wanted`? If so, I would suggest mentioning about it in `Manager`'s fields documentation